consensus/safety-rules/src/safety_rules.rs

```diff
@@ -0,0 +1,248 @@
+    // that a new proposal's parent is higher or equal to the preferred_round.
+    preferred_round: Round,
+}
+
+impl Display for ConsensusState {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "ConsensusState: [\n\
+             \tepoch = {},
+             \tlast_voted_round = {},\n\
+             \tpreferred_round = {}\n\
+             ]",
+            self.epoch, self.last_voted_round, self.preferred_round
+        )
+    }
+}
+
+impl ConsensusState {
+    pub fn new(epoch: u64, last_voted_round: Round, preferred_round: Round) -> Self {
+        Self {
+            epoch,
+            last_voted_round,
+            preferred_round,
+        }
+    }
+
+    /// Returns the current epoch
+    pub fn epoch(&self) -> u64 {
+        self.epoch
+    }
+
+    /// Returns the last round that was voted on
+    pub fn last_voted_round(&self) -> Round {
+        self.last_voted_round
+    }
+
+    /// Returns the preferred block round
+    pub fn preferred_round(&self) -> Round {
+        self.preferred_round
+    }
+}
+
+/// SafetyRules is responsible for two things that are critical for the safety of the consensus:
+/// 1) voting rules,
+/// 2) commit rules.
+/// SafetyRules is NOT THREAD SAFE (should be protected outside via e.g., RwLock).
+/// The commit decisions are returned to the caller as result of learning about a new QuorumCert.
+pub struct SafetyRules {
+    persistent_storage: Box<dyn PersistentStorage>,
+    validator_signer: Arc<ValidatorSigner>,
+}
+
+impl SafetyRules {
+    /// Constructs a new instance of SafetyRules given the BlockTree and ConsensusState.
+    pub fn new(
+        persistent_storage: Box<dyn PersistentStorage>,
+        validator_signer: Arc<ValidatorSigner>,
+    ) -> Self {
+        Self {
+            persistent_storage,
+            validator_signer,
+        }
+    }
+
+    pub fn signer(&self) -> &ValidatorSigner {
+        &self.validator_signer
+    }
+
+    /// Learn about a new quorum certificate. Several things can happen as a result of that:
+    /// 1) update the preferred block to a higher value.
+    /// 2) commit some blocks.
+    /// In case of commits the last committed block is returned.
+    /// Requires that all the ancestors of the block are available for at least up to the last
+    /// committed block, might panic otherwise.
+    /// The update function is invoked whenever a system learns about a potentially high QC.
+    pub fn update(&mut self, qc: &QuorumCert) {
+        if qc.ledger_info().ledger_info().epoch() > self.persistent_storage.epoch() {
+            self.persistent_storage
+                .set_epoch(qc.ledger_info().ledger_info().epoch());
+            self.persistent_storage.set_last_voted_round(0);
+            self.persistent_storage.set_preferred_round(0);
+        } else if qc.parent_block().round() > self.persistent_storage.preferred_round() {
```

Comment:
hmm, sorry if my previous comment was misleading, this should not be an else, it should alway go through.

if we have a QC{epoch: 2, parent_block_round: 10} and locally we have {epoch: 1, parent_block_round: 100}, we need not only update epoch but also update preferred round to be 10 too to avoid violating safety rules.

we need some test cases for reconfiguration in safety rules, but that's on me

Reply:
Hmm... that seems.... Ideally we would see a proof of epoch change and cause the change. I guess that is todo :). With that in mind, I'm going to make another PR littered with TODOs as a means to coordinate.

Reply:
sure, we need TODO, although the proof to change epoch is orthogonal to the preferred round update after epoch change there?